### DISCUSSION TITLE
Language setup

### DISCUSSION THREAD

**Discussion by kobzar** [LazyVim/LazyVim#1081] [Q&A]
Hi. I'm just a new user and need help with language settings mechanism.
I tried to find samples here and on the home page, but couldn't find any.
Only ideas and ways hot to make it. But it must be easy for my opinion.
So, i ask help with configurations and hope that this post help many others peoples.

Lets play with simple and standard situation. We are coding web cite on python + web framework like django. 
So we are working with **.py .css .html .js** files.

**1. LSP server**
   We are installed a few LSP servers for python in Mason:
    pyright
    pylyzer
    ruff
    How to choosee needed from the list? I dont want to run all of them in onetime.
    (something like **...python.lsp_server = {"pyright"}** or **...python.lsp_server = {"pyright", "ruff"}** )

**2. Linters**
    The same question as first but for linters. 
    We are installed a few linters  for python in Mason:
    flake8
    ruff
    pylint
    How to select needed for python files?
    How to add rules for selected linter (Ignore [E501] ... e.t.c.)

    
**3. Formating** (The hardest question)
As i say before we are working with different file types, so must use different formatters.
For example:
  black - for python files
  prettier - for html files
  stylua - for lua files
And we are must add formating parameter for each of them (line length e.t.c.)

For another nvim build like LunarVim this actions has good documentation and samples 
https://www.lunarvim.org/docs/configuration/language-features/linting-and-formatting
So all of my question look easy
```lua
-- Disable one and use another LSP (both of them is installed)
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "pyright" })
require("lspconfig").pylyzer.setup({})

-- Configure formatters with parameters
local formatters = require("lvim.lsp.null-ls.formatters")
formatters.setup({
	{ command = "black", extra_args = { "--line-length", "120" }, filetypes = { "python" } },
	{ command = "isort", filetypes = { "python" } },
	{ command = "stylua", filetypes = { "lua" } },
	{ command = "rustfmt", filetypes = { "rust" } },
})

-- Configure inter with parameters
local linters = require("lvim.lsp.null-ls.linters")
linters.setup({
	{ command = "flake8", extra_args = { "--max-line-length", "120" }, filetypes = { "python" } },
	{ command = "luacheck", extra_args = { "--ignore", "112" }, filetypes = { "lua" } },
})
```
I undestand that this strings just make some changes in isp configuration files, so my question
how to make the same in LazyVim?

In documentation i find just this 
```lua
{
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      pyright = {},
      ruff_lsp = {},
    },
  },
  setup = {
    ruff_lsp = function()
      require("lazyvim.util").on_attach(function(client, _)
        if client.name == "ruff_lsp" then
          -- Disable hover in favor of Pyright
          client.server_capabilities.hoverProvider = false
        end
      end)
    end,
  },
}
```
And this isn't anough for undestanding. Just add pylyzer = {}, in server sections and this sample no more sense. 

Thans for all who will  help and attach their configuration.
(One more time: its must be easy and understanded for user like in any other editor)

**Comment by dpetka2001** [comment]
There's an extra in LazyVim as of a few days ago, that provides all this functionality for python without any need for manual setup. You just have to put the following in `lua/config/lazy.lua`
```lua
require("lazy").setup({
  spec = {
    { "LazyVim/LazyVim", import = "lazyvim.plugins" },
    { import = "lazyvim.plugins.extras.lang.python" },
    { import = "plugins" },
  },
})
```

**Reply by dpetka2001** [reply]
Just to take any kind of confusion out of the way, just paste the line about python in that file because the rest of the lines already exist, so you avoid any kind of duplicate entries.

**Comment by kobzar** [comment]
I was added this line and dont make any another changes in my configuration about python or LSP 
And now when i open the LspInfo
```
 Language client log: /home/kobzar/.local/state/nvim/lsp.log
 Detected filetype:   python
 
 2 client(s) attached to this buffer: 
 
 Client: ruff_lsp (id: 1, bufnr: [27])
 	filetypes:       python
 	autostart:       true
 	root directory:  Running in single file mode.
 	cmd:             /home/kobzar/.local/share/nvim/mason/bin/ruff-lsp
 
 Client: pyright (id: 2, bufnr: [27])
 	filetypes:       python
 	autostart:       true
 	root directory:  Running in single file mode.
 	cmd:             /home/kobzar/.local/share/nvim/mason/bin/pyright-langserver --stdio
 
 Configured servers list: bashls, pyright, ruff_lsp, jsonls, lua_ls, rust_analyzer, taplo
```

But, i dont like use ruff_lsp. Formating isnt work. 
How to enable black with parameters, flake8, isort and chage pyright to pylyzer with the same settings for example?

**Reply by dpetka2001** [reply]
Oh yes, they removed the formatters from the extras because there were quite controversial opinions around them. You can use the example from [Go](https://www.lazyvim.org/plugins/extras/lang.go#null-lsnvim) to setup Python then. Just put the sources that you want to use instead of the ones that Go uses and keep the rest the same. Also keep in mind that the maintainer of `null-ls`, which is used quite a lot for formatters and linters archived the repo and currently we are waiting to see if there will be a continuation (by someone forking it) or something else entirely or some of the other existing solutions step up.

**Comment by kobzar** [comment]
Maybe my english is worse but i ask again
I am need use black for python files, stylua for lua files and prettier for html only files. If i will add 
something like 
```nls.builtins.formatting.prettier``` - it will use fo all files
And the same question as before - hot to add option to formatter like 
```{ command = "black", extra_args = { "--line-length", "120" }, filetypes = { "python" } },```

**Comment by dpetka2001** [comment]
Have you read the [null-ls docs](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/BUILTIN_CONFIG.md)? You can configure which filetypes you want with something like
```lua
null_ls.builtins.formatting.prettier.with({
        filetypes = { "html", "json", "yaml", "markdown" },
    })
```
For extra arguments you can pass them with 
```lua
null_ls.builtins.formatting.shfmt.with({
        extra_args = { "-i", "2", "-ci" },
    }),
```
Again, plz make the effort to read the docs. Most of the things you're asking are described in the docs.

**Comment by GabeSoler** [comment]
I find it pretty sad that the answer is 'read the docs' when the docs of something 'lazy' are not at all easy to comprehend. I just went trying to understand what to do to use Django in lazy vim, and the Python section does not make it clear at all how to change those configurations. 
Sadly Django does not seem to work well with those :/

**Comment by Jacobus-afk** [comment]
i can't really comment on the quality of the lazy docs, except to say try to welcome the discomfort of having to figure things out.. because it really does get easier to tailor your config once you've built up a mental model of how lazy does things.

however, abstractions on top of abstractions arent good some times, and it might be better to just use vanilla neovim to do a config anyways nowadays (especially for nvim v0.12)

anyways, here's my lazyvim config for django dev, hope it'll give you a starting point

`lua/config/options.lua`

```lua
-- LSP Server to use for Python.
-- Set to "basedpyright" to use basedpyright instead of pyright.
vim.g.lazyvim_python_lsp = "basedpyright"
```

`lua/plugins/lspconfig`

```lua
return {
	"neovim/nvim-lspconfig",
	opts = function(_, opts)

		opts.servers.ruff = {
			init_options = {
				settings = {
					lineLength = 110,
				},
			},
			capabilities = {
				general = {
					-- neovim 0.11 fix
					-- https://github.com/astral-sh/ruff/issues/14483
					positionEncodings = { "utf-16" },
				},
			},
		}

		opts.servers.basedpyright = {
			settings = {
				basedpyright = {
					analysis = {
						typeCheckingMode = "off",
						ignore = { "*" },
						autoSearchPaths = true,
						useLibraryCodeForTypes = true,
						diagnosticMode = "openFilesOnly",
					},
					disableOrganizeImports = true,
				},
			},
		}
	end,
}
```

`lua/plugins/conform.lua`

```lua
return {
	{
		"stevearc/conform.nvim",
		opts = {
			formatters_by_ft = {
				python = { "ruff_organize_imports", lsp_format = "first" },
				htmldjango = { "djhtml" },
			},
			formatters = {
				ruff_organize_imports = {
					append_args = { "--line-length", "110" },
				},
				djhtml = {
					inherit = false,
					command = "djhtml",
					args = {
						"$FILENAME",
						"--tabwidth",
						"2",
					},
					stdin = false,
				},
			},
		},
	},
}
```

remember to have the a virtual environment activated for full code navigation and completion

im not sure if you need the lazyvim python extra.. if i remember correctly mason will autoinstall ruff and basedpyright for you. djhtml (for django templates formatting) you might have to install manually with mason
